Chatkit/Auth/TokenProvider.swift

```diff
@@ -0,0 +1,137 @@
+public typealias TokenProvider = PPTokenProvider
+public typealias TokenProviderResult = PPTokenProviderResult
+
+/// A factory for `TokenProvider`s.
+///
+/// The methods on this object create token provider instances which can interface with an HTTP endpoint in your backend, or the Chatkit Test Token Provider (for use during development)
+public struct ChatkitTokenProviders {
+    /// Creates a `TokenProvider` which fetches tokens from the Chatkit Test Token Provider. It should be used for development purposes only.
+    ///
+    /// The Chatkit Test Token Provider is a token provider endpoint hosted by Pusher which returns tokens for the requested `userId` without authenticating the user. It is intended for development convenience and *must not be enabled in instances which host production data*.
+    ///
+    /// The test token provider must be enabled for your instance in the Chatkit dashboard.
+    ///
+    /// - Parameters:
+    ///     - instanceLocator: The locator string for your instance, which you also provide when initialising the `Chatkit` object.
+    ///     - userId: The user identifier to fetch a token for. The provider will sign a valid token for any user identifier requested, without applying any kind of authentication.
+    /// - Returns:a `TokenProvider` which fetches tokens from the Chatkit Test Token Provider.
+    public static func chatkitTestTokenProvider(instanceLocator: String, userId: String) -> TokenProvider {
+        return CachingTokenProvider(
+            delegateProvider: RetryingTokenProvider(
+                delegateProvider: testTokenHTTPCallProvider(
+                    instanceLocator: instanceLocator, userId: userId
+                ),
+                retryAttempts: 3,
+                retryDelayMs: 1000
+            )
+        )
+    }
+
+    /// Create a `TokenProvider` which fetches tokens from an HTTP endpoint in your backend, with caching and retry implemented. This is the recommended implementation for production use.
+    ///
+    /// - Parameters:
+    ///     - method: The HTTP method to use in the call, e.g. `GET`, `POST`
+    ///     - host: The host name to be called, e.g. `example.com`
+    ///     - path: The path component of the URL to be called, e.g. `/tokens`
+    ///     - headers: An optional map of headers to include in the request. Here you can supply the session or other credientials which your endpoint might require to authenticate the request.
+    ///     - queryParams: An optional map of query parameters to include in the request URL. Here you can supply the session or other credientials which your endpoint might require to authenticate the request.
+    /// - Returns: a `TokenProvider` which fetches tokens from the specified HTTP endpoint, and is wrapped in default caching and retry logic.
+    public static func httpTokenProvider(method: String, host: String, path: String, headers: [String: String]?, queryParams: [String: String]?) -> TokenProvider {
```

Comment:
Wouldn't it be good to have `POST` as default? It's about creating an access session so probably POST fits it better than GET?

I reckon HTTPs will be used for the communication. Wouldn't it be worth to convey it at least in the docs? Now someone may be worried that it's not encrypted HTTP.

Reply:
As a default for the `method` parameter? We could do, but I'm not sure it's worth much to the developer (who still has to check that method their backend exposes)

Reply:
They need to check but most likely it will be POST. Also they won't need to remember to type it in such case.

Reply:
What happens when client's session expires or is invalidated? Do they need to reinstantiate the whole Chatkit just to create a new TokenProvider? Will there'll be an option to set a new TokenProvider? If yes, probably it would be better to have `TokenProviderFactory` which takes the data that does not change as constructor params and then have a factory method like `createTokenProviderWithRequestHeader(name: String, value: String)` (excuse my Kotlin;) + other less convenient/popular factory methods for multiple headers or changing query params.

Reply:
When the token expires, the provider fetches a new one.

The most naive provider fetches a token on every call, which is why we wrap it in a `CachingTokenProvider` which re-uses a token until it expires.

Reply:
That means that the SDK gets 401 from the BE and calls fetch token on the token provider?

Reply:
I've checked and it looks like currently fetchToken is called on every request. Still my question above stays valid on what shall happen when token expires or is invalidated on the backend side but the local token provider is not aware of it. 

Still if fetch doesn't need to mean actual fetch, wouldn't it be better to name it as retrieval function, e.g. `getToken`?

Reply:
> Still my question above stays valid on what shall happen when token expires or is invalidated on the backend side but the local token provider is not aware of it.

Yes, if the token is rejected by the backend, then a new one should be requested.

> Still if fetch doesn't need to mean actual fetch, wouldn't it be better to name it as retrieval function, e.g. getToken?

Aside: The convention on iOS is that the `get` prefix actually denotes expensive actions, not property access.

But to answer your question, it's always a fetch, but sometimes it's cached. It's always asynchronous, and it could hit the network for any given call, it's just that a lot of the time it's been optimised away. To me that means the name should reflect the "worst case".

Reply:
That's fine, on Android get with async interface will also convey that the operation may take longer. I think that we should abstract from calling it "fetch" if it's not always fetch. Indeed if we suggest caching and avoiding additional RTT hitting a separate host before every communication with Chatkit...

Alternative names could be: `provideToken` or `retrieveToken`. They also abstract from if it's local memory access or a fetch taking longer time. However simple `getToken` feels like the best and the most common name.

Now it may be understood that TokenProvider interface function is called "fetch" as we do caching outside of it. Some integrators may be surprised to see much higher hit rates than they expected for their token provider endpoint.

Reply:
Regarding the TokenProvider interface I would like to also propose better naming of `clearToken` function: `refreshToken` which would indicate that the next call to `getToken` should return a fresh token. That is actually the trigger to call `clearToken`. It's being called when we get 401 passed down by our backend. Also currently having the token param there seems to be unnecessary complexity.

`clearToken` feels more like clearing it for security purposes when one stops using Chatkit. It feels even more weird when called in the context of an immutable String (which cannot be cleared) and moreover having another instance of immutable sting with that token content passed as a param. If we want to expose TokenProvider implementation which does caching then it should have a function `clearCachedToken` but it should not be a part of the generic TokenProvider interface.

It appears thought that iOS version of TokenProvider does not have `clearToken` function, at least I don't see it implemented in `HTTPSTokenProvider`. Is there a difference between declaration of TokenProvider on Android and iOS in the platform lib?

Reply:
This has likely changed around a bit since I simplified and made the caching and retry an internal concern. Perhaps we can talk about this in person.